### DISCUSSION TITLE
Package Manager - v2

### DISCUSSION THREAD

**Discussion by mawildoer** [atopile/atopile#495] [Ideas]
It's time for v2 of [packages.atopile.io](https://packages.atopile.io)

## Objective:
- 🚀 Make finding and reusing designs fast and easy!
- 🚧 Make a very low bar to sharing and contributing a design
- 🤗 Make it rewarding to contribute and get recognition
- 🧠 Encourage exploration and experimentation

Having used the v1 package manager for roughly a year now, we've learned a bit about how we use it, and what's missing.

Compared to software packages, we've noticed a few unique characteristics of packaging electronics:
- Much smaller package sizes: Even a 50-line description of a component is useful!
- Fewer updates: For many or most packages, there isn't much to keep up with! Chips are typically stable for years if not decades

## Concept for v2

- Treat packages likes components: select them based on an abstract spot in your design
- Parametric search for packages by whatever attributes they have
- Make sharing the default and easiest option, and quality improves with usage.
- Clear indication of a package's quality and activity level

### Finding packages can happen two ways

1. The compiler finds a package that matches an abstraction in your design, and uses it for you
2. You search for and find it explicitly on packages.atopile.io

### How sharing a package could work

1. Jerry makes a useful module in some OS software he's working on.
2. Jerry pushes his project to a public and MIT licensed repo.
3. atopile indexes Jerry's public project and finds a useful looking module within it!
4. atopile creates a `packagelet` from the module, and it's dependencies by analyzing Jerry's project.
5. atopile automatically publishes the `packagelet` to packages.atopile.io
   The `packagelet` is now available for others to use!
   It's clearly marked as a `packagelet` (eg. quality unknown) under Jerry's Github username, and points back to Jerry's repo.
6. People start using the `packagelet` in their projects!
7. After some excitement and usage 🚀, atopile asks Jerry if he'd be interested in turning the `packagelet` into a fully fledged package
8. Jerry agrees 🤝, and uses an "ejection tool" atopile provides to elevate the `packagelet` to a `package` owned by him.
   The `package` now appears elevated in search results with a shiny new badge, and is happily used by the community.

```mermaid
flowchart TD
    A[Project in Public Repo] -->|atopile indexes| B[Useful Module Detected]
    B -->|Auto-publish| C[packagelet on packages.atopile.io]
    C -->|Owner Ejects Package| D[Fully-fledged package on packages.atopile.io]
```

## Soliciting!

- Packages: Get ready to share!
- Ideas!
- What should we call them? Rust has crates, Python has packages, Julia has modules, etc.

**Comment by mawildoer** [comment]
Some other thoughts and "mood board ideas"

I really like the npm bar graph of quality indicators:
![Screenshot 2024-10-31 at 15 52 54](https://github.com/user-attachments/assets/4018b280-3fb3-4d15-8458-81032d1bbe8c)

PyPi's detail's sidebar is all I've wanted there
![Screenshot 2024-10-31 at 16 23 15](https://github.com/user-attachments/assets/bc5f3386-ace3-43e8-85e3-2f3ffe3fc039)

**Reply by rubenbaldewsing** [reply]
+1 for some sort of quality check and display.

**Comment by sethfischer** [comment]
### Decouple package name from repository name

e.g. a package named `koa-speer-rc-series` has its source in a repository named `example.com/user/atopile-koa-speer-rc-series`.

**Comment by rubenbaldewsing** [comment]
Make https://packages.atopile.io/ more mobile friendly.

**Comment by rubenbaldewsing** [comment]
Automatically add links to documents/documentation on detail page.

**Comment by rubenbaldewsing** [comment]
Index and link to projects that are using the module.

**Comment by mawildoer** [comment]
Here's an example of a packagelet's content:
[code-as-data-demo.zip](https://github.com/user-attachments/files/17940175/code-as-data-demo.zip)

**Reply by rubenbaldewsing** [reply]
Maybe making an additional reference design mandatory or making 2 types of packagelets (1 with other without ref design). You will barely use "raw" components in a product design, especially if the reference designs are configurable.

**Comment by rubenbaldewsing** [comment]
ESPHome has similar support for _external components_. For inspiration [docs](https://esphome.io/components/external_components)

**Comment by rubenbaldewsing** [comment]
Add an MCP server connected to packages.atopile.io so you can just  chat your projects together in cursor (or other LLM)
